backend/uclapi/oauth/views.py

```diff
@@ -0,0 +1,347 @@
+            "error": "The signed data received failed the signature check. Either the server did not respond in a timely manner, or the data was tampered with."
+        })
+
+    # Since the data has passed verification at this point, and we have checked the validity of the client secret, we can
+    # now generate an OAuth access token for the user.
+    # But first, we should check if a token has been generated already.
+    # If a token does already exist then we should not add yet another one to the database. We can just pass those keys to the app
+    # again (in case it has lost them).
+
+    try:
+        token = OAuthToken.objects.get(app=app, user=user)
+
+        # If the code gets here then the user has used this app before, so let's check that the scope does
+        # not need changing
+        if not token.scope.scopeIsEqual(app.scope):
+            # Remove the current scope from the token
+            token.scope.delete()
+
+            # Clone the scope of the app
+            app_scope = app.scope
+            app_scope.id = None
+            app_scope.save()
+
+            # Assign the new scope to the token
+            token.scope = app_scope
+
+            # Save the token with the new scope
+            token.save()
+
+    except OAuthToken.DoesNotExist:
+        # The user has never logged in before so let's clone the scope and create a brand new
+        # OAuth token
+
+        # Clone the scope defined in the app model
+        app_scope = app.scope
+        app_scope.id = None
+        app_scope.save()
+
+        # Now set up a new token with that scope
+        token = OAuthToken(
+            app=app,
+            user=user,
+            scope=app_scope
+        )
+        token.save()
+
+    # Now that we have a token we can pass one back to the app
+    # We'll make a final HTTP request to the app and provide the state and the OAuth token.
+    # We sincerely hope they'll save this token!
+    # The app can use the token to pull in any personal data (name, UPI, etc.) later on, so
+    # we won't bother to give it to them just yet.
+
+    oauth_data = {
+        "state": state,
+        "client_id": app.client_id,
+        "token": token.token,
+        "scope": json.dumps(token.scope.scopeDict())
+    }
+
+    # Now forward them the OAuth token for the user. If they're not happy with that then
+    # that's their fault after the final redirect!
+    oauth_req = requests.post(app.callback_url + "/token", data=oauth_data)
+
+    # Now redirect the user back to the app, at long last.
+    # Just in case they've tried to be super clever and host multiple apps with the same
+    # callback URL, we'll provide the client ID along with the state
+    return redirect(app.callback_url + "?client_id=" + app.client_id + "&state=" + state)
+
+def userdata(request):
+    try:
+        token_code = request.GET.get("token")
+        token = OAuthToken.objects.get(token=token_code)
+    except:
+        return JsonResponse({
```

Comment:
I'm guessing this response should be a more general one that future apps/services can also make use of. Probably a decorator?